File: forms/outpatient-outcome-report/front-end-form-with-svelte/src/lib/engine/prom-domain.ts
import type { AssessmentData, DomainGrade, FiredRule } from './types';
import { eq5dSummary, promisGphTScore, promisMhTScore } from './utils';

/**
 * Grade an individual instrument from its change direction.
 *
 * Returns:
 *   'improved' | 'worsened' | 'stable' | 'missing'
 */
type InstrumentResult = 'improved' | 'worsened' | 'stable' | 'missing';

/** Grade the EQ-5D-5L instrument (dimensions + VAS combined). */
function gradeEq5d(data: AssessmentData): InstrumentResult {
	const prom = data.promEq5d5l;
	const { improved, worsened, missing } = eq5dSummary(prom);
	const total = 6; // 5 dimensions + VAS
	if (missing === total) return 'missing';
	if (worsened > improved) return 'worsened';
	if (improved > worsened && improved > 0) return 'improved';
	return 'stable';
}

/** Grade the GRC instrument. */
function gradeGrc(data: AssessmentData): InstrumentResult {
	const grc = data.promGrc.globalRatingOfChange;
	if (grc === null) return 'missing';
	if (grc > 0) return 'improved';
	if (grc < 0) return 'worsened';
	return 'stable';
}

/** Grade the PROMIS Global Health instrument by mapping the linear T-score
 * approximation to a tri-state outcome.
 *
 * The approximation in `utils.ts` spans roughly 16.2–67.7 with midpoint near 42
 * (item value 3 across the relevant questions). Thresholds are calibrated to
 * that midpoint rather than the population norm of 50:
 *   T ≥ 50 → improved
 *   T 35–49 → stable
 *   T < 35 → worsened
 *   null → missing
 *
 * Production deployments must use the official IRT calibration tables and
 * the population-norm thresholds (≥55/≥45/<45) instead.
 */
function gradePromis(data: AssessmentData): InstrumentResult {
	const p = data.promPromis;
	const gph = promisGphTScore(p);
	const gmh = promisMhTScore(p);
	if (gph === null && gmh === null) return 'missing';

	const scores = [gph, gmh].filter((s): s is number => s !== null);
	const avg = scores.reduce((a, b) => a + b, 0) / scores.length;

	if (avg >= 50) return 'improved';
	if (avg >= 35) return 'stable';
	return 'worsened';
}

/**
 * Grade the PROM domain from EQ-5D-5L, GRC, and PROMIS composite.
 *
 * Rules (from spec):
 * - ≥ 2 instruments improved → A or B (A if all 3 improved, B if 2 improved)
 * - All stable (or 1 improved, rest stable) → C
 * - Any instrument worsened ≥ 1 level → D
 * - Multiple worsened → E
 * - All missing → '' (no grade)
 */
export function gradePROM(data: AssessmentData): { grade: DomainGrade; rules: FiredRule[] } {
	const results: { name: string; result: InstrumentResult }[] = [
		{ name: 'EQ-5D-5L', result: gradeEq5d(data) },
		{ name: 'GRC', result: gradeGrc(data) },
		{ name: 'PROMIS', result: gradePromis(data) }
	];

	const rules: FiredRule[] = [];
	const nonMissing = results.filter((r) => r.result !== 'missing');

	if (nonMissing.length === 0) {
		return { grade: '', rules };
	}

	const improvedCount = nonMissing.filter((r) => r.result === 'improved').length;
	const worsenedCount = nonMissing.filter((r) => r.result === 'worsened').length;

	let grade: DomainGrade;

	if (worsenedCount >= 2) {
		grade = 'E';
		rules.push({
			id: 'PROM-005',
			domain: 'PROM',
			description: 'Multiple PROM instruments worsened',
			grade: 'E'
		});
	} else if (worsenedCount === 1) {
		grade = 'D';
		const worsened = nonMissing.find((r) => r.result === 'worsened')!;
		rules.push({
			id: 'PROM-004',
			domain: 'PROM',
			description: `${worsened.name} shows worsening`,
			grade: 'D'
		});
	} else if (improvedCount >= 3) {
		grade = 'A';
		rules.push({
			id: 'PROM-001',
			domain: 'PROM',
			description: 'All three PROM instruments improved',
			grade: 'A'
		});
	} else if (improvedCount >= 2) {
		grade = 'B';
		rules.push({
			id: 'PROM-002',
			domain: 'PROM',
			description: 'Two or more PROM instruments improved',
			grade: 'B'
		});
	} else {
		grade = 'C';
		rules.push({
			id: 'PROM-003',
			domain: 'PROM',
			description: 'PROM instruments stable (no significant improvement or worsening)',
			grade: 'C'
		});
	}

	return { grade, rules };
}

File: forms/outpatient-outcome/full-stack-with-loco-tera-htmx-alpine/src/engine/prom_domain.rs
use super::types::{AssessmentData, DomainGrade, FiredRule};
use super::utils::{eq5d_summary, promis_gph_t_score, promis_mh_t_score};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum InstrumentResult {
    Improved,
    Worsened,
    Stable,
    Missing,
}

fn grade_eq5d(data: &AssessmentData) -> InstrumentResult {
    let s = eq5d_summary(&data.prom_eq5d5l);
    let total = 6; // 5 dimensions + VAS
    if s.missing == total {
        return InstrumentResult::Missing;
    }
    if s.worsened > s.improved {
        return InstrumentResult::Worsened;
    }
    if s.improved > s.worsened && s.improved > 0 {
        return InstrumentResult::Improved;
    }
    InstrumentResult::Stable
}

fn grade_grc(data: &AssessmentData) -> InstrumentResult {
    match data.prom_grc.global_rating_of_change {
        None => InstrumentResult::Missing,
        Some(v) if v > 0 => InstrumentResult::Improved,
        Some(v) if v < 0 => InstrumentResult::Worsened,
        Some(_) => InstrumentResult::Stable,
    }
}

fn grade_promis(data: &AssessmentData) -> InstrumentResult {
    let gph = promis_gph_t_score(&data.prom_promis);
    let gmh = promis_mh_t_score(&data.prom_promis);
    if gph.is_none() && gmh.is_none() {
        return InstrumentResult::Missing;
    }
    let scores: Vec<f64> = [gph, gmh].iter().filter_map(|s| *s).collect();
    let avg: f64 = scores.iter().sum::<f64>() / scores.len() as f64;
    if avg >= 50.0 {
        InstrumentResult::Improved
    } else if avg >= 35.0 {
        InstrumentResult::Stable
    } else {
        InstrumentResult::Worsened
    }
}

pub struct PromResult {
    pub grade: DomainGrade,
    pub rules: Vec<FiredRule>,
}

/// Grade the PROM domain from EQ-5D-5L, GRC, and PROMIS composite.
pub fn grade_prom(data: &AssessmentData) -> PromResult {
    let results: [(&str, InstrumentResult); 3] = [
        ("EQ-5D-5L", grade_eq5d(data)),
        ("GRC", grade_grc(data)),
        ("PROMIS", grade_promis(data)),
    ];

    let mut rules: Vec<FiredRule> = Vec::new();
    let non_missing: Vec<&(&str, InstrumentResult)> = results
        .iter()
        .filter(|(_, r)| *r != InstrumentResult::Missing)
        .collect();

    if non_missing.is_empty() {
        return PromResult { grade: String::new(), rules };
    }

    let improved_count = non_missing.iter().filter(|(_, r)| *r == InstrumentResult::Improved).count();
    let worsened_count = non_missing.iter().filter(|(_, r)| *r == InstrumentResult::Worsened).count();

    let grade: DomainGrade = if worsened_count >= 2 {
        rules.push(FiredRule {
            id: "PROM-005".to_string(),
            domain: "PROM".to_string(),
            description: "Multiple PROM instruments worsened".to_string(),
            grade: "E".to_string(),
        });
        "E".to_string()
    } else if worsened_count == 1 {
        let worsened_name = non_missing
            .iter()
            .find(|(_, r)| *r == InstrumentResult::Worsened)
            .map(|(n, _)| *n)
            .unwrap_or("");
        rules.push(FiredRule {
            id: "PROM-004".to_string(),
            domain: "PROM".to_string(),
            description: format!("{worsened_name} shows worsening"),
            grade: "D".to_string(),
        });
        "D".to_string()
    } else if improved_count >= 3 {
        rules.push(FiredRule {
            id: "PROM-001".to_string(),
            domain: "PROM".to_string(),
            description: "All three PROM instruments improved".to_string(),
            grade: "A".to_string(),
        });
        "A".to_string()
    } else if improved_count >= 2 {
        rules.push(FiredRule {
            id: "PROM-002".to_string(),
            domain: "PROM".to_string(),
            description: "Two or more PROM instruments improved".to_string(),
            grade: "B".to_string(),
        });
        "B".to_string()
    } else {
        rules.push(FiredRule {
            id: "PROM-003".to_string(),
            domain: "PROM".to_string(),
            description: "PROM instruments stable (no significant improvement or worsening)"
                .to_string(),
            grade: "C".to_string(),
        });
        "C".to_string()
    };

    PromResult { grade, rules }
}
